routers/web/web.go

```diff
@@ -1210,13 +1210,15 @@ func RegisterRoutes(m *web.Route) {
 				m.PostOptions("/git-receive-pack", repo.ServiceReceivePack)
 				m.GetOptions("/info/refs", repo.GetInfoRefs)
 				m.GetOptions("/HEAD", repo.GetTextFile("HEAD"))
+				m.GetOptions("/config", repo.GetTextFile("config"))  // needed by git-annex's dumb http mode
 				m.GetOptions("/objects/info/alternates", repo.GetTextFile("objects/info/alternates"))
 				m.GetOptions("/objects/info/http-alternates", repo.GetTextFile("objects/info/http-alternates"))
 				m.GetOptions("/objects/info/packs", repo.GetInfoPacks)
 				m.GetOptions("/objects/info/{file:[^/]*}", repo.GetTextFile(""))
 				m.GetOptions("/objects/{head:[0-9a-f]{2}}/{hash:[0-9a-f]{38}}", repo.GetLooseObject)
 				m.GetOptions("/objects/pack/pack-{file:[0-9a-f]{40}}.pack", repo.GetPackFile)
 				m.GetOptions("/objects/pack/pack-{file:[0-9a-f]{40}}.idx", repo.GetIdxFile)
+				m.GetOptions("/annex/objects/{hash1}/{hash2}/{keyDir}/{key}", repo.GetAnnexObject) // for git-annex
```

Comment:
I wonder if we could do this?

```suggestion
				if !setting.Annex.Enabled { 
				 	m.GetOptions("/config", repo.GetTextFile("config"))
					m.GetOptions("/annex/objects/{hash1}/{hash2}/{keyDir}/{key}", repo.GetAnnexObject)
				}
```

Reply:
> ignSignInAndCsrf seems to only have an effect on POST requests, which these routes aren't ([link](https://github.com/neuropoly/gitea/blob/2ec990931f61a1ef670b88c3bf37b06d4550548d/routers/web/web.go#L236) and [link](https://github.com/neuropoly/gitea/blob/2ec990931f61a1ef670b88c3bf37b06d4550548d/modules/context/auth.go#L65))

That's...a good point. But I don't trust I understand all the implications here, so I'm going to fall onto the heuristic of copying what the similar routes do. The LFS routes and git-http routes have it attached, even though some of them are GET requests. If it doesn't apply, then it should just be a harmless no-op. But if it should apply, it's probably pretty important that it's there.

Perhaps one day we will convince joeyh to support annex uploads over HTTP, and on that day we'll need to add a POST route and this will apply?

> UserAssignmentWeb ([link](https://github.com/neuropoly/gitea/blob/2ec990931f61a1ef670b88c3bf37b06d4550548d/services/context/user.go#L16)) has something to do with getting the username from request parameters (and doing some validation? not sure about the logic), which we don't use or require at the moment (although maybe something to explore for access controls)

I tried taking out UserAssignmentWeb and got a panic. It's needed by `httpBase()` to enforce permissions:

<details>

```
                    testlogger.go:78: 2022/09/19 11:53:24 routers/web/base.go:135:1() [E] [63289074-46] PANIC: runtime error: invalid memory address or nil pointer dereference
                        	/usr/lib/go-1.18/src/runtime/panic.go:220 (0x457855)
                        		panicmem: panic(memoryError)
                        	/usr/lib/go-1.18/src/runtime/signal_unix.go:818 (0x457825)
                        		sigpanic: panicmem()
                        	/home/GRAMES.POLYMTL.CA/p115628/src/neurogitea/gitea/models/user/user.go:451 (0x1ff73e2)
                        		(*User).IsOrganization: return u.Type == UserTypeOrganization
                        	/home/GRAMES.POLYMTL.CA/p115628/src/neurogitea/gitea/routers/web/repo/http.go:114 (0x1ff73ea)
                        		httpBase: if !owner.IsOrganization() && !owner.IsActive {
                        	/home/GRAMES.POLYMTL.CA/p115628/src/neurogitea/gitea/routers/web/repo/http.go:565 (0x1ffc673)
                        		GetTextFile.func1: h := httpBase(ctx)
                        	/home/GRAMES.POLYMTL.CA/p115628/src/neurogitea/gitea/modules/web/wrap_convert.go:47 (0x1bd9e56)
                        		convertHandler.func3: t(ctx)
                        	/home/GRAMES.POLYMTL.CA/p115628/src/neurogitea/gitea/modules/web/wrap.go:41 (0x1bd8089)
                        		wrapInternal.func1: done, deferrable := handler(resp, req, others...)
                        	/usr/lib/go-1.18/src/net/http/server.go:2084 (0x981c8e)
                        		HandlerFunc.ServeHTTP: f(w, r)
                        	/home/GRAMES.POLYMTL.CA/p115628/go/pkg/mod/github.com/go-chi/chi/v5@v5.0.7/mux.go:442 (0x1ba5bf5)
                        		(*Mux).routeHTTP: h.ServeHTTP(w, r)
                        	/usr/lib/go-1.18/src/net/http/server.go:2084 (0x981c8e)
                        		HandlerFunc.ServeHTTP: f(w, r)
                        	/home/GRAMES.POLYMTL.CA/p115628/src/neurogitea/gitea/modules/web/wrap.go:63 (0x1bd850f)
                        		Middle.func1.1: next.ServeHTTP(ctx.Resp, ctx.Req)
                        	/usr/lib/go-1.18/src/net/http/server.go:2084 (0x981c8e)
                        		HandlerFunc.ServeHTTP: f(w, r)
                        	/home/GRAMES.POLYMTL.CA/p115628/src/neurogitea/gitea/modules/web/wrap.go:63 (0x1bd850f)
                        		Middle.func1.1: next.ServeHTTP(ctx.Resp, ctx.Req)
                        	/usr/lib/go-1.18/src/net/http/server.go:2084 (0x981c8e)
                        		HandlerFunc.ServeHTTP: f(w, r)
                        	/home/GRAMES.POLYMTL.CA/p115628/src/neurogitea/gitea/modules/web/wrap.go:63 (0x1bd850f)
                        		Middle.func1.1: next.ServeHTTP(ctx.Resp, ctx.Req)
                        	/usr/lib/go-1.18/src/net/http/server.go:2084 (0x981c8e)
                        		HandlerFunc.ServeHTTP: f(w, r)
                        	/home/GRAMES.POLYMTL.CA/p115628/go/pkg/mod/github.com/go-chi/chi/v5@v5.0.7/middleware/get_head.go:37 (0x1e98e04)
                        		GetHead.func1: next.ServeHTTP(w, r)
                        	/usr/lib/go-1.18/src/net/http/server.go:2084 (0x981c8e)
                        		HandlerFunc.ServeHTTP: f(w, r)
                        	/home/GRAMES.POLYMTL.CA/p115628/src/neurogitea/gitea/modules/web/wrap.go:63 (0x1bd850f)
                        		Middle.func1.1: next.ServeHTTP(ctx.Resp, ctx.Req)
                        	/usr/lib/go-1.18/src/net/http/server.go:2084 (0x981c8e)
                        		HandlerFunc.ServeHTTP: f(w, r)
                        	/home/GRAMES.POLYMTL.CA/p115628/src/neurogitea/gitea/modules/context/context.go:786 (0x1bba75a)
                        		Contexter.func1.1: next.ServeHTTP(ctx.Resp, ctx.Req)
                        	/usr/lib/go-1.18/src/net/http/server.go:2084 (0x981c8e)
                        		HandlerFunc.ServeHTTP: f(w, r)
                        	/home/GRAMES.POLYMTL.CA/p115628/go/pkg/mod/github.com/!n!y!times/gziphandler@v1.1.1/gzip.go:342 (0x2077ef0)
                        		GzipHandlerWithOpts.func1.1: h.ServeHTTP(w, r)
                        	/usr/lib/go-1.18/src/net/http/server.go:2084 (0x981c8e)
                        		HandlerFunc.ServeHTTP: f(w, r)
                        	/home/GRAMES.POLYMTL.CA/p115628/go/pkg/mod/github.com/go-chi/chi/v5@v5.0.7/mux.go:71 (0x1ba3a2c)
                        		(*Mux).ServeHTTP: mx.handler.ServeHTTP(w, r)
                        	/home/GRAMES.POLYMTL.CA/p115628/go/pkg/mod/github.com/go-chi/chi/v5@v5.0.7/mux.go:314 (0x1ba53db)
                        		(*Mux).Mount.func1: handler.ServeHTTP(w, r)
                        	/usr/lib/go-1.18/src/net/http/server.go:2084 (0x981c8e)
                        		HandlerFunc.ServeHTTP: f(w, r)
                        	/home/GRAMES.POLYMTL.CA/p115628/go/pkg/mod/github.com/go-chi/chi/v5@v5.0.7/mux.go:442 (0x1ba5bf5)
                        		(*Mux).routeHTTP: h.ServeHTTP(w, r)
                        	/usr/lib/go-1.18/src/net/http/server.go:2084 (0x981c8e)
                        		HandlerFunc.ServeHTTP: f(w, r)
                        	/home/GRAMES.POLYMTL.CA/p115628/src/neurogitea/gitea/routers/web/base.go:175 (0x207a9f1)
                        		Recovery.func1.1: next.ServeHTTP(w, req)
                        	/usr/lib/go-1.18/src/net/http/server.go:2084 (0x981c8e)
                        		HandlerFunc.ServeHTTP: f(w, r)
                        	/home/GRAMES.POLYMTL.CA/p115628/go/pkg/mod/gitea.com/go-chi/session@v0.0.0-20211218221615-e3605d8b28b8/session.go:257 (0x1ae743d)
                        		Sessioner.func1.1: next.ServeHTTP(w, req)
                        	/usr/lib/go-1.18/src/net/http/server.go:2084 (0x981c8e)
                        		HandlerFunc.ServeHTTP: f(w, r)
                        	/home/GRAMES.POLYMTL.CA/p115628/src/neurogitea/gitea/modules/web/wrap.go:110 (0x1bd8fa8)
                        		WrapWithPrefix.func1.1: next.ServeHTTP(resp, req)
                        	/usr/lib/go-1.18/src/net/http/server.go:2084 (0x981c8e)
                        		HandlerFunc.ServeHTTP: f(w, r)
                        	/home/GRAMES.POLYMTL.CA/p115628/go/pkg/mod/github.com/go-chi/chi/v5@v5.0.7/mux.go:71 (0x1ba3a2c)
                        		(*Mux).ServeHTTP: mx.handler.ServeHTTP(w, r)
                        	/home/GRAMES.POLYMTL.CA/p115628/go/pkg/mod/github.com/go-chi/chi/v5@v5.0.7/mux.go:314 (0x1ba53db)
                        		(*Mux).Mount.func1: handler.ServeHTTP(w, r)
                        	/usr/lib/go-1.18/src/net/http/server.go:2084 (0x981c8e)
                        		HandlerFunc.ServeHTTP: f(w, r)
                        	/home/GRAMES.POLYMTL.CA/p115628/go/pkg/mod/github.com/go-chi/chi/v5@v5.0.7/mux.go:442 (0x1ba5bf5)
                        		(*Mux).routeHTTP: h.ServeHTTP(w, r)
                        	/usr/lib/go-1.18/src/net/http/server.go:2084 (0x981c8e)
                        		HandlerFunc.ServeHTTP: f(w, r)
                        	/home/GRAMES.POLYMTL.CA/p115628/src/neurogitea/gitea/routers/common/middleware.go:79 (0x1e9e402)
                        		Middlewares.func2.1: next.ServeHTTP(resp, req)
                        	/usr/lib/go-1.18/src/net/http/server.go:2084 (0x981c8e)
                        		HandlerFunc.ServeHTTP: f(w, r)
                        	/home/GRAMES.POLYMTL.CA/p115628/src/neurogitea/gitea/modules/web/routing/logger_manager.go:123 (0x1bd3daf)
                        		(*requestRecordsManager).handler.func1: next.ServeHTTP(w, req)
                        	/usr/lib/go-1.18/src/net/http/server.go:2084 (0x981c8e)
                        		HandlerFunc.ServeHTTP: f(w, r)
                        	/home/GRAMES.POLYMTL.CA/p115628/go/pkg/mod/github.com/go-chi/chi/v5@v5.0.7/middleware/strip.go:30 (0x1e9bb78)
                        		StripSlashes.func1: next.ServeHTTP(w, r)
                        	/usr/lib/go-1.18/src/net/http/server.go:2084 (0x981c8e)
                        		HandlerFunc.ServeHTTP: f(w, r)
                        	/home/GRAMES.POLYMTL.CA/p115628/go/pkg/mod/github.com/chi-middleware/proxy@v1.1.1/middleware.go:37 (0x1e7bc36)
                        		ForwardedHeaders.func1.1: h.ServeHTTP(w, r)
                        	/usr/lib/go-1.18/src/net/http/server.go:2084 (0x981c8e)
                        		HandlerFunc.ServeHTTP: f(w, r)
                        	/home/GRAMES.POLYMTL.CA/p115628/src/neurogitea/gitea/routers/common/middleware.go:32 (0x1e9e252)
                        		Middlewares.func1.1: next.ServeHTTP(context.NewResponse(resp), req.WithContext(ctx))
                        	/usr/lib/go-1.18/src/net/http/server.go:2084 (0x981c8e)
                        		HandlerFunc.ServeHTTP: f(w, r)
                        	/home/GRAMES.POLYMTL.CA/p115628/go/pkg/mod/github.com/go-chi/chi/v5@v5.0.7/mux.go:88 (0x1ba39e1)
                        		(*Mux).ServeHTTP: mx.handler.ServeHTTP(w, r)
                        	/home/GRAMES.POLYMTL.CA/p115628/src/neurogitea/gitea/modules/web/route.go:200 (0x1bd74cd)
                        		(*Route).ServeHTTP: r.R.ServeHTTP(w, req)
                        	/usr/lib/go-1.18/src/net/http/server.go:2916 (0x98527a)
                        		serverHandler.ServeHTTP: handler.ServeHTTP(rw, req)
                        	/usr/lib/go-1.18/src/net/http/server.go:1966 (0x980736)
                        		(*conn).serve: serverHandler{c.server}.ServeHTTP(w, w.req)
                        	/usr/lib/go-1.18/src/runtime/asm_amd64.s:1571 (0x474940)
                        		goexit: BYTE	$0x90	// NOP
```

</details>

I didn't write any access control in -- I called `httpBase()` like the other functions in routers/web/repo/http.go and it enforced them for me.

Reply:
Like I said below though, I did apply the `annexEnabled` middleware suggestion :)